server/src/main/java/com/linecorp/centraldogma/server/storage/encryption/DefaultEncryptionStorageManager.java

```diff
@@ -0,0 +1,418 @@
+                            nonce = Arrays.copyOfRange(metadataValue, 0, NONCE_SIZE_BYTES);
+                            keyInDefaultColumnFamily = EncryptionKeyUtil.encrypt(dek, nonce, idPart);
+                        } else {
+                            logger.warn("Invalid metadata value for object key: {}",
+                                        new String(metadataKey, StandardCharsets.UTF_8));
+                        }
+                    } else {
+                        logger.warn("Invalid object metadata key length: {}",
+                                    new String(metadataKey, StandardCharsets.UTF_8));
+                    }
+                } else if (startsWith(metadataKey, rev2ShaPrefixBytes)) {
+                    // MetadataKey: project/repo/rev2sha/<revision_major_bytes(4)>
+                    // MetadataValue: nonce(12)
+                    if (metadataKey.length == rev2ShaPrefixBytes.length + 4) {
+                        idPart = Arrays.copyOfRange(metadataKey, rev2ShaPrefixBytes.length, metadataKey.length);
+                        nonce = metadataValue;
+                        if (nonce != null && nonce.length == NONCE_SIZE_BYTES) {
+                            keyInDefaultColumnFamily = EncryptionKeyUtil.encrypt(dek, nonce, idPart);
+                        } else {
+                            logger.warn("Invalid nonce (metadata value) for rev2sha key: {}",
+                                        new String(metadataKey, StandardCharsets.UTF_8));
+                        }
+                    } else {
+                        logger.warn("Invalid rev2sha metadata key length: {}",
+                                    new String(metadataKey, StandardCharsets.UTF_8));
+                    }
+                } else if (startsWith(metadataKey, refsKeyPrefixBytes)) {
+                    // MetadataKey: project/repo/refs/... (e.g., project/repo/refs/heads/master)
+                    // idPart should be "refs/..." so copy from projectRepoPrefixBytes.length
+                    // MetadataValue: nonce(12)
+                    idPart = Arrays.copyOfRange(metadataKey, projectRepoPrefixBytes.length, metadataKey.length);
+                    nonce = metadataValue;
+                    if (nonce != null && nonce.length == NONCE_SIZE_BYTES) {
+                        keyInDefaultColumnFamily = EncryptionKeyUtil.encrypt(dek, nonce, idPart);
+                    } else {
+                        logger.warn("Invalid nonce (metadata value) for ref key: {}",
+                                    new String(metadataKey, StandardCharsets.UTF_8));
+                    }
+                } else {
+                    logger.warn("Unknown metadata key pattern for prefix {}: {}",
+                                projectRepoPrefix, new String(metadataKey, StandardCharsets.UTF_8));
+                }
+
+                writeBatch.delete(metadataColumnFamilyHandle, metadataKey);
+                deletedCount++;
```

Comment:
Some repositories may have 1M+ commits. Would you test to delete a large repository and measure the followings:
- How long the removal takes.
- How much memory is consumed. I'd like to know if all delete operations are cached in the RocksDB memory for atomicity.

Reply:
Addressed here:
https://github.com/line/centraldogma/pull/1119/commits/910a9835eba906aa3c05e0ca949f6b8b8fcd520b